### ISSUE TITLE
Performant DetailsList + CommandBar

### ISSUE THREAD

**Issue by pedro-beslogic** [OfficeDev/office-ui-fabric-react#12123]
#### Describe the feature that you would like added

A more performant DetailsList and `CommandBar`
#### What component or utility would this be added to

DetailsList and `CommandBar`

#### Additional context/screenshots

OBS: Since I was not sure if this was an issue or a feature request, I have also posted the below text on [stackoverflow](https://stackoverflow.com/questions/60455657/how-to-make-detailslist-more-performant-when-rendering-a-custom-component-in-the
) with the #office-ui-fabric tag, as recommended

I'm trying to render a not so big DetailsList (something between 100-200 rows) where I have a custom column with the CommandBarcomponent. One of the actions is a delete action that deletes the row itself. The problem that I am facing with this is that after I click to delete the row, the code takes more time that I would expect to rerender, producing the feel of a slow system.

Running in dev, I get a pretty slow performance that is quite unpleasant. (My machine is a i7 with 16GbRam so the problem should not be my machine).

Running in production though I get an 'ok' performance but not a great performance as I would like.

Code sample: https://codepen.io/pedro-beslogic/pen/eYNWLLo

I have commented the line 141 to exemplify what I consider a great performance

React profiler file: https://beslogic-my.sharepoint.com/:u:/p/pedro_ferreira/EQ-P0K6-uYZFqhPNIcMtbmsBNWU6CUXrQjuexCPhPkibpQ?e=JMjnoa

Chrome performance file: https://beslogic-my.sharepoint.com/:u:/p/pedro_ferreira/EbogBtQwPWVCpwCrIIFfAdYBsOX1BlYB0cOWl7Zy1KwanA?e=Z53Qyi

Am I doing something wrong or is it a better way to make the table more performant?

**Comment by aneeshack4** [comment]
Hi @pedro-beslogic thanks for reaching out to us! And for providing this codepen. Are you saying that this delete is causing slower perf - I just wanted to clarify since the comment says "fast perf":

>  //FAST PERFORMANCE
> // <h1 onClick={() => dispatch({ type: "delete", payload: { id: data.id } })}>DELETE</h1>

Not sure why there would be a diff in perf between prod and dev... maybe the variables in your dev environment are set differently - have you checked those?

**Comment by pedro-beslogic** [comment]
> Hi @pedro-beslogic thanks for reaching out to us! And for providing this codepen. Are you saying that this delete is causing slower perf - I just wanted to clarify since the comment says "fast perf":
> 
> > //FAST PERFORMANCE
> > // <h1 onClick={() => dispatch({ type: "delete", payload: { id: data.id } })}>DELETE
> 
> Not sure why there would be a diff in perf between prod and dev... maybe the variables in your dev environment are set differently - have you checked those?

Yes, when rendering the table with a light "Custom component" which in this example is
`<h1 onClick={() => dispatch({ type: "delete", payload: { id: data.id } })}>DELETE</h1>`(The fast one to render)  the performance is great. But when I use the `CommandBar` as the custom component, (the slow one to render, at line `113`)  I get a poor performance. I understand that the CommandBar is heavier to render than a simple header with a text but I believe that it's too slow since I am not rendering so much data (100-200 rows).  About the difference between prod and dev, I believe that the difference rely more on [react ](https://reactjs.org/docs/optimizing-performance.html) than on ui-fabric. But even using the production mode, the performance is better but still not great, so it's something about the combination the `CommandBar`, the `DetailsList` or even both that makes the interaction to be slow.

**Comment by aneeshack4** [comment]
@pedro-beslogic Thanks for explaining this further. So it sounds like the problem is CommandBar needs to be more performant... I do know that we plan to have a big focus on making the next version of all our components including these two much more performant and we're taking steps towards that but it'll take a while. To make these two performant right away doesn't seem doable because as far as I understand, they'd be pretty involved fixes. Correct me if I'm wrong - tagging component owners @joschect @micahgodbolt

**Comment by micahgodbolt** [comment]
Command bar is 100% not the right control for your use case. CommandBar does a TON of expensive logic to resize iteself to fit the given space. If you're only using it for the overflow, you can do the exact same thing with a IconButton with MenuProps. You'll have dramatically better perf.

**Comment by pedro-beslogic** [comment]
Indeed, the performance was dramatically improved with the use of IconButton, thank you very much for the support. 
Yet, I had to deduce the use of the IconButton manually because the current [documentation](https://developer.microsoft.com/en-us/fabric#/controls/web) does not have any information on this. I found a microsoft [link](https://docs.microsoft.com/en-us/javascript/api/office-ui-fabric-react/iconbutton?view=office-ui-fabric-react-latest) that poorly describes the usage of the IconButton. 

I believe that could be interesting to describe the IconButton in the https://developer.microsoft.com/en-us/fabric#/controls page and even maybe to illustrate the usage of IconButton in the examples, or at least suggest to avoid the usage of CommandBar inside the DetailList in the **Do's and Dont's** section
